presto-main/src/main/java/com/facebook/presto/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -0,0 +1,190 @@
+        List<Type> inputTypes = ImmutableList.of(type);
+        MethodHandle inputFunction;
+        MethodHandle outputFunction;
+        Class<? extends AccumulatorState> stateInterface = BigIntegerAndLongState.class;
+        AccumulatorStateSerializer<?> stateSerializer = new BigIntegerAndLongStateSerializer();
+
+        if (((DecimalType) type).isShort()) {
+            inputFunction = SHORT_DECIMAL_INPUT_FUNCTION;
+            outputFunction = SHORT_DECIMAL_OUTPUT_FUNCTION;
+        }
+        else {
+            inputFunction = LONG_DECIMAL_INPUT_FUNCTION;
+            outputFunction = LONG_DECIMAL_OUTPUT_FUNCTION;
+        }
+        inputFunction = inputFunction.bindTo(type);
+        outputFunction = outputFunction.bindTo(type);
+
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(NAME, type.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
+                createInputParameterMetadata(type),
+                inputFunction,
+                null,
+                null,
+                COMBINE_FUNCTION,
+                outputFunction,
+                stateInterface,
+                stateSerializer,
+                new BigIntegerAndLongStateFactory(),
+                type,
+                false);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        GenericAccumulatorFactoryBinder factory = new AccumulatorCompiler().generateAccumulatorFactoryBinder(metadata, classLoader);
+        return new InternalAggregationFunction(NAME, inputTypes, intermediateType, type, true, false, factory);
+    }
+
+    private static List<ParameterMetadata> createInputParameterMetadata(Type type)
+    {
+        return ImmutableList.of(new ParameterMetadata(STATE), new ParameterMetadata(BLOCK_INPUT_CHANNEL, type), new ParameterMetadata(BLOCK_INDEX));
+    }
+
+    public static void inputShortDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        accumulateValueInState(BigInteger.valueOf(type.getLong(block, position)), state);
+    }
+
+    public static void inputLongDecimal(Type type, BigIntegerAndLongState state, Block block, int position)
+    {
+        accumulateValueInState(decodeUnscaledValue(type.getSlice(block, position)), state);
+    }
+
+    private static void accumulateValueInState(BigInteger value, BigIntegerAndLongState state)
+    {
+        initializeIfNeeded(state);
+        state.setBigInteger(state.getBigInteger().add(value));
+        state.setLong(state.getLong() + 1);
+    }
+
+    private static void initializeIfNeeded(BigIntegerAndLongState state)
+    {
+        if (state.getBigInteger() == null) {
+            state.setBigInteger(BigInteger.valueOf(0));
+        }
+    }
+
+    public static void combine(BigIntegerAndLongState state, BigIntegerAndLongState otherState)
+    {
+        state.setLong(state.getLong() + otherState.getLong());
```

Comment:
@haozhun why couldn't we change this after merging in our changes? As you see other classes behave similarly to this one (you change them in your PR), as I see this PR does not breaking anything, it's working within what is established today. Then after we land this we could safely change the assumptions about the State objects that you don't like in the framework. We've been rebasing this PR for some time now as time goes by, so I think if this is not a show-stopper we should follow the chronological order here. Thanks.